File: app/src/main/java/org/neidhardt/dynamicsoundboard/misc/SoundProgressAdapter.java
package org.neidhardt.dynamicsoundboard.misc;

import android.os.Handler;
import android.support.v7.widget.RecyclerView;
import org.neidhardt.dynamicsoundboard.mediaplayer.EnhancedMediaPlayer;
import org.neidhardt.dynamicsoundboard.soundmanagement.ServiceManagerFragment;

import java.util.ArrayList;
import java.util.List;

public abstract class SoundProgressAdapter<T extends RecyclerView.ViewHolder>
		extends
			RecyclerView.Adapter<T>
		implements
			Runnable
{
	public static final int UPDATE_INTERVAL = 500;
	protected final Handler handler = new Handler();

	protected ServiceManagerFragment serviceManagerFragment;

	public void setServiceManagerFragment(ServiceManagerFragment serviceManagerFragment)
	{
		this.serviceManagerFragment = serviceManagerFragment;
	}

	/**
	 * Starts periodic updates of sounds loaded in the adapter. This is used to update the progress bars of running sounds.
	 */
	public void startProgressUpdateTimer()
	{
		this.handler.postDelayed(this, UPDATE_INTERVAL);
	}

	public void stopProgressUpdateTimer()
	{
		this.handler.removeMessages(0);
	}

	@Override
	public void run()
	{
		List<Integer> itemsWithProgressChanged = getPlayingItems();
		if (itemsWithProgressChanged == null || itemsWithProgressChanged.size() == 0)
		{
			stopProgressUpdateTimer();
			return;
		}
		for (Integer index : itemsWithProgressChanged)
			notifyItemChanged(index);

		this.startProgressUpdateTimer();
	}

	private List<Integer> getPlayingItems()
	{
		List<Integer> playingSounds = new ArrayList<>();
		List<EnhancedMediaPlayer> allSounds = this.getValues();
		int count = allSounds.size();
		for (int i = 0; i < count; i++)
		{
			if (allSounds.get(i).isPlaying())
				playingSounds.add(i);
		}
		return playingSounds;
	}

	protected abstract List<EnhancedMediaPlayer> getValues();
}

File: app/src/main/kotlin/org/neidhardt/dynamicsoundboard/views/recyclerviewhelpers/SoundProgressAdapter.kt
package org.neidhardt.dynamicsoundboard.views.recyclerviewhelpers

import android.os.Handler
import android.support.v7.widget.RecyclerView
import org.neidhardt.dynamicsoundboard.mediaplayer.MediaPlayerController
import org.neidhardt.ui_utils.recyclerview.adapter.BaseAdapter
import java.util.*
import java.util.concurrent.atomic.AtomicBoolean

val UPDATE_INTERVAL: Int = 500

abstract class SoundProgressAdapter<T>
(
		private val  recyclerView: RecyclerView
) :
		BaseAdapter<MediaPlayerController, T>(),
		SoundProgressTimer,
		Runnable
where T : RecyclerView.ViewHolder, T : SoundProgressViewHolder
{
	private val handler: Handler = Handler()
	private val hasTimerStarted: AtomicBoolean = AtomicBoolean(false)

	/**
	 * Starts periodic updates of sounds loaded in the adapter. This is used to update the progress bars of running sounds.
	 */
	override fun startProgressUpdateTimer()
	{
		if (!this.hasTimerStarted.getAndSet(true))
			this.handler.postDelayed(this, UPDATE_INTERVAL.toLong())
	}

	override fun stopProgressUpdateTimer()
	{
		if (this.hasTimerStarted.getAndSet(false))
			this.handler.removeCallbacks(this)
	}

	override fun run()
	{
		val itemsWithProgressChanged = getPlayingItems()
		if (itemsWithProgressChanged.size == 0)
		{
			this.stopProgressUpdateTimer()
			return
		}
		for (index in itemsWithProgressChanged)
		{
			val viewHolderToUpdate = this.recyclerView.findViewHolderForAdapterPosition(index) as SoundProgressViewHolder?
			viewHolderToUpdate?.onProgressUpdate()
		}
		this.hasTimerStarted.set(false)
		this.startProgressUpdateTimer()
	}

	private fun getPlayingItems(): List<Int>
	{
		val playingSounds = ArrayList<Int>()
		val allSounds = this.values
		val count = allSounds.size
		for (i in 0..count - 1)
		{
			if (allSounds[i].isPlayingSound)
				playingSounds.add(i)
		}
		return playingSounds
	}

}
